core/string/char_utils.h

```diff
@@ -41,7 +41,7 @@
 	int middle;                                          \
                                                          \
 	while (low <= high) {                                \
-		middle = (low + high) / 2;                       \
+		middle = low + ((high - low) >> 1);              \
```

Comment:
`>> 1` instead of `/ 2` feels unnecessary, it's harder to read and compilers already do this optimization

Reply:
The compiler does not optimize this. https://godbolt.org/z/WEa7vT6PG

Reply:
That's because it's unsafe to shift negative numbers

Reply:
> That's because it's unsafe to shift negative numbers

Compiler does this optimization instead: http://www.flounder.com/multiplicative_inverse.htm

Reply:
Indeed, `a >> 1` is rounded to negative infinity, but `a / 2` is rounded to zero, it doesn't matter in this case of course, so right shift is safe (even considering the implementation defined behavior)

Reply:
Yes, but `low` and `high` are always positive. How do you feel about using `uint32_t` instead of `int`?

Reply:
It returns an `int` so unsure if that'd be appropriate, I don't think there's any issue with using shifts here, as the difference should always be non-negative (the signedness of `low` and `high` is less of an issue here)

Reply:
> I don't think there's any issue with using shifts here

Well, they seem to be. I compiled a binary search using `int` and `uint_32`.

int
```
.LBB0_7:                                # =>This Inner Loop Header: Depth=1
        mov     esi, ecx
        sub     esi, edx
        sar     esi
        add     esi, edx
        movsxd  r8, esi
        mov     r8d, dword ptr [rdi + 8*r8]
        cmp     r8d, 111
        jge     .LBB0_8
        cmp     r8d, 110
        je      .LBB0_10
        inc     esi
        mov     edx, esi
        cmp     edx, ecx
        jle     .LBB0_7
```

uint
```
.LBB0_6:                                # =>This Inner Loop Header: Depth=1
        mov     edx, eax
        sub     edx, ecx
        shr     edx
        add     edx, ecx
        mov     esi, dword ptr [rdi + 8*rdx]
        cmp     esi, 111
        jge     .LBB0_7
        cmp     esi, 110
        je      .LBB0_12
        inc     edx
        mov     ecx, edx
        cmp     ecx, eax
        jbe     .LBB0_6
```
`movsxd  r8, esi` this instructions when using the `int`

Reply:
Weirdly but `clang` doesn't seem to recognize that a certain value is always positive and doesn't perform one shift optimization and `MSVC` adds some instructions for some reason, unlike `g++`. That's an issue with `clang` and `MSVC`, but we could just change `int` to `unsigned int` when appropriate, that helps `clang` and `MSVC` to optimize. [g++](https://godbolt.org/z/xPGvh1z1q), [MSVC with int](https://godbolt.org/z/b5Psa7af8), [MSVC with unsigned int](https://godbolt.org/z/reEfYz7aG), [clang with int](https://godbolt.org/z/n5bMW1TY4), [clang with unsigned int](https://godbolt.org/z/GhW94jvcT).

Reply:
You could also simply do:
```suggestion
		middle = low + ((unsigned int)(high - low) / 2); \
```
Try and see

Reply:
> Try and see

Works with [clang](https://godbolt.org/z/38ETd6xY3) and [MSVC](https://godbolt.org/z/d11fTn9Mc), on `MSVC` it even produces identical code, before there were paddings, but that's just generally an issue with `MSVC` in a lot of places.

Reply:
I just thought of something. It seems to be safe to use from `uint` to `int` if the `uint` is not very large. And because of this, we can leave it as it was here `middle = (low + high) / 2`, but use `uint`. This way there will be no overflow due to an extra bit. And if there is an overflow using `uint`, then it will be even for `middle = low + ((high - low) / 2)`;

Reply:
This will reduce the number of operations by 2 times. And it will be able to work with more data (from 1 billion to 2 billion)

Reply:
How will it work with more data? The index is a signed integer still, and the indices are always non-negative, so the difference always fits

Reply:
> This will reduce the number of operations by 2 times.

It works! [g++](https://godbolt.org/z/GxbqcEzGG), [clang](https://godbolt.org/z/dsa4G4qPb), [MSVC](https://godbolt.org/z/seG33aEvc).

Reply:
My PR simply solves the problem `(low + high)` when `low`, `high` approach the maximum int (2 billion). In reality, if you look at these bits after addition, they are the same. But the point is how `/2` will behave with them. If we have signed, it will divide incorrectly. For example, `0|100 + 0|100 = 1|00` (overflow due to which it will divide incorrectly)

Reply:
> How will it work with more data?

0100+0100 = 1000
1000 / 2 = 0100
This is how it will work with larger data.

Reply:
> the difference always fits (the sum however doesn't)

The sum will fit if cast from `int` to `unsigned int` before adding numbers together. Basically what Nazarwadim is proposing. Though I agree that this won't allow working with more data than the current approach by the PR.

Reply:
Oh you meant by solving this bug, it sounded like you meant more data than already supported :) I assumed you meant beyond just fixing the overflow

Reply:
> Indeed, I know

Just got a litle bit confused then.

Reply:
I just meant for the indices themselves, not the operation, i.e. that using unsigned integers won't change the range we can work with (assuming an algorithm that takes things into account)